### PR TITLE
Eventbus-system

### PR DESCRIPTION
**Pull request by jakkeez**
A system to publish and subscribe to events throughout the project. 
Two test classes are implemented into sandbox main to test the subscribe and publish feature with a simple text event. TODO before merge: Remove these after rest of pull request has been approved.

### CODE REVIEW

File: StortSpelProjekt/Engine/src/Events/EventBus.h [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	HandlerList* handlers = m_Subscribers[typeid(EventType)];
+
+	//No handler exists, exit
+	if (handlers == nullptr)
+	{
+		return;
+	}
+
+	//Loop through each handler and execute events
+	for (auto& handler : *handlers) {
```
Comment (LeoWikstrom) [inline StortSpelProjekt/Engine/src/Events/EventBus.h]: Curly bracket on next line

File: StortSpelProjekt/Engine/src/Events/EventBus.h [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+		if (handler != nullptr) {
```
Comment (LeoWikstrom) [inline StortSpelProjekt/Engine/src/Events/EventBus.h]: Curly bracket on next line

File: StortSpelProjekt/Engine/src/Events/EventBus.h [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+#include <typeindex>
+#include "MemberFunctionWrapper.h"
+#include "Events.h"
+
+typedef std::list<HandlerFunctionBase*> HandlerList;
+class EventBus 
+{
+public:
+	template<typename EventType>
+	void publish(EventType* evnt);
```
Comment (LeoWikstrom) [inline StortSpelProjekt/Engine/src/Events/EventBus.h]: Start with capital letter

File: StortSpelProjekt/Engine/src/Events/EventBus.h [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+	template<class T, class EventType>
+	void subscribe(T* instance, void (T::* memberFunction)(EventType*));
```
Comment (LeoWikstrom) [inline StortSpelProjekt/Engine/src/Events/EventBus.h]: Capital letter

File: StortSpelProjekt/Engine/src/Events/EventBus.h [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+	//Singleton of the eventbus
+	static EventBus& get();
```
Comment (LeoWikstrom) [inline StortSpelProjekt/Engine/src/Events/EventBus.h]: Capital letter. Recommend renaming to GetInstance(), to match with the other singleton classes in the project (Option and Input)

File: StortSpelProjekt/Engine/src/Events/MemberFunctionWrapper.h [hunk 6]
```diff
@@ -0,0 +1,37 @@
+#pragma once
+#include "Events.h"
+
+// This is the interface for MemberFunctionWrapper that each specialization will use
+class HandlerFunctionBase 
+{
+public:
+    // Call the member function
+    void exec(Event* evnt)
```
Comment (LeoWikstrom) [inline StortSpelProjekt/Engine/src/Events/MemberFunctionWrapper.h]: Capital letter

File: StortSpelProjekt/Engine/src/Events/MemberFunctionWrapper.h [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    MemberFunctionHandler(T* instance, MemberFunction memberFunction) : instance{ instance }, memberFunction{ memberFunction } {};
+
+    void call(Event* evnt) 
+    {
+        // Cast event to the correct type and call member function
+        (instance->*memberFunction)(static_cast<EventType*>(evnt));
+    }
+private:
+    // Pointer to class instance
+    T* instance;
```
Comment (LeoWikstrom) [inline StortSpelProjekt/Engine/src/Events/MemberFunctionWrapper.h]: m_pInstance

File: StortSpelProjekt/Engine/src/Events/MemberFunctionWrapper.h [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+    // Pointer to member function
+    MemberFunction memberFunction;
```
Comment (LeoWikstrom) [inline StortSpelProjekt/Engine/src/Events/MemberFunctionWrapper.h]: m_MemberFunction

**Review by LeoWikstrom — CHANGES_REQUESTED** [review]

**Comment by jakkeez** [comment]
https://medium.com/@savas/nomad-game-engine-part-7-the-event-system-45a809ccb68f
Documentation for the eventbus

**Review by LeoWikstrom — DISMISSED** [review]

**Review by Schtekt — CHANGES_REQUESTED** [review]
Looks good! Please remove the tests before the merge commences.

**Review by Schtekt — APPROVED** [review]

**Review by LeoWikstrom — APPROVED** [review]

**Pull request merged**